diff --git a/src/3rdparty/webkit/Source/WebCore/features.pri b/src/3rdparty/webkit/Source/WebCore/features.pri
index f04d0b4..5e4bfbe 100644
--- a/src/3rdparty/webkit/Source/WebCore/features.pri
+++ b/src/3rdparty/webkit/Source/WebCore/features.pri
@@ -163,10 +163,10 @@ symbian|maemo5|maemo6 {
     DEFINES += ENABLE_VIDEO=0
 
     mac:!contains(DEFINES, USE_QTMULTIMEDIA=1) {
-        DEFINES -= ENABLE_VIDEO=0
-        DEFINES += ENABLE_VIDEO=1
-        DEFINES += WTF_USE_QTKIT=1
-        DEFINES -= WTF_USE_QTKIT=0
+        # DEFINES -= ENABLE_VIDEO=0
+        # DEFINES += ENABLE_VIDEO=1
+        # DEFINES += WTF_USE_QTKIT=1
+        # DEFINES -= WTF_USE_QTKIT=0
     } else: linux-*:!contains(DEFINES, USE_QTMULTIMEDIA=1) {
         !contains(QT_CONFIG, no-pkg-config):system(pkg-config --exists glib-2.0 gio-2.0 gstreamer-0.10): {
             DEFINES -= ENABLE_VIDEO=0
