diff --git a/src/3rdparty/webkit/Source/WebCore/rendering/RenderFileUploadControl.cpp b/src/3rdparty/webkit/Source/WebCore/rendering/RenderFileUploadControl.cpp
index 87f6ff1..7d8f2d1 100644
--- a/src/3rdparty/webkit/Source/WebCore/rendering/RenderFileUploadControl.cpp
+++ b/src/3rdparty/webkit/Source/WebCore/rendering/RenderFileUploadControl.cpp
@@ -130,9 +130,6 @@ void RenderFileUploadControl::chooseIconForFiles(FileChooser* chooser, const Vec
 
 void RenderFileUploadControl::click()
 {
-    // Requires a user gesture to open the file dialog.
-    if (!frame() || !frame()->loader()->isProcessingUserGesture())
-        return;
     if (Chrome* chromePointer = chrome())
         chromePointer->runOpenPanel(frame(), m_fileChooser);
 }
